File: rudder-client-ios/RudderClient/EventRepository.swift
//
//  RudderEventRepository.swift
//  RudderSample
//
//  Created by Arnab Pal on 04/07/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Alamofire
import Foundation

internal class EventRepository {
    private var eventBuffer = [RudderEvent]()
    private var endPointUri: String = Constants.END_POINT_URL
    private var flushQueueSize: Int = 10
    static var eventTemplate: RudderEventTemplate? = nil
    
    func initiate(endPointUri: String, flushQueueSize: Int) {
        self.endPointUri = endPointUri
        self.flushQueueSize = flushQueueSize
     
        EventRepository.eventTemplate = RudderEventTemplate.populate()
    }
    
    func getFlushQueueSize() -> Int {
        return self.flushQueueSize
    }
 
    func setFlushQueueSize(flushQueueSize: Int) {
        if flushQueueSize < self.flushQueueSize {
            flush()
        }
        
        self.flushQueueSize = flushQueueSize
    }
    
    func getEndPointUri() -> String {
        return self.endPointUri
    }
    
    func setEndPointUri(endPointUri: String) {
        self.endPointUri = endPointUri
    }
    
    func dump(event: RudderEvent) {
        if (Constants.DEBUG) {
            do {
                try print("EVENT: " , String(bytes: JSONEncoder().encode(event), encoding: .utf8) ?? "no event")
            } catch {
                // ignored
            }
        }
        
        eventBuffer.append(event)
        
        if (eventBuffer.count == self.flushQueueSize) {
            self.flushEvents()
        }
    }
    
    func flush() {
        self.flushEvents()
    }
    
    private func flushEvents() {
        let payload = RudderEventPayload(batch: eventBuffer)
        
        do {
            let url = endPointUri+"/hello"
            if (Constants.DEBUG) {
                print("URL: " + url)
            }
            
            Alamofire.request(url, method: .post, parameters: try payload.toDictionary() as Parameters, encoding: JSONEncoding.prettyPrinted, headers: nil)
                .responseString { dataResponse in
                    print("RESPONSE: " + dataResponse.description)
                 
                    if (dataResponse.error != nil) {
                        print("ERROR: " + dataResponse.error!.localizedDescription)
                    }
                    
                    self.eventBuffer.removeAll()
                }
        } catch {
            print("ERROR: " + error.localizedDescription)
        }
    }
}
    
extension DateFormatter {
    static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd-MM-yyyy HH:mm:ssZ"
        return formatter
    }()
}

File: rudder-client-unity/sdk/rudder-unity-sdk/EventRepository.cs
using System;
using com.rudderlabs.unity.library.Event;
using System.Collections.Generic;
using Newtonsoft.Json;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace com.rudderlabs.unity.library
{
    internal class EventRepository
    {
        private int flushQueueSize = Constants.FLUSH_QUEUE_SIZE;
        private string endPointUri = Constants.BASE_URL;
        private List<RudderEvent> eventBuffer = new List<RudderEvent>();

        internal EventRepository(int flushQueueSize, string endPointUri)
        {
            this.endPointUri = endPointUri;
            this.flushQueueSize = flushQueueSize;
        }

        internal void Dump(RudderEvent rudderEvent)
        {
            eventBuffer.Add(rudderEvent);

            if (eventBuffer.Count == flushQueueSize)
            {
                FlushEventsAsync();
            }
        }

        internal void FlushEventsAsync()
        {
            RudderEventPayload eventPayload = new RudderEventPayload(eventBuffer);
            string payloadString = JsonConvert.SerializeObject(eventPayload);
            Console.Write("~~REQUEST: " + payloadString);

            Task<HttpResponseMessage> task = PostEventToServer(payloadString);

            
        }

        private async Task<HttpResponseMessage> PostEventToServer(string payload)
        {
            using (var client = new HttpClient())
            {
                using (var request = new HttpRequestMessage(HttpMethod.Post, endPointUri+ "/hello"))
                {
                    request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                    request.Content = new StringContent(payload, Encoding.UTF8, "application/json");
                    return await client.SendAsync(request);
                }
            }
        }

        internal int GetFlushQueueSize()
        {
            return flushQueueSize;
        }

        internal string GetEndPointUri()
        {
            return endPointUri;
        }

        internal void SetFlushQueueSize(int _flushQueueSize)
        {
            flushQueueSize = _flushQueueSize;
        }

        internal void SetEndPointUri(string _endPointUri)
        {
            endPointUri = _endPointUri;
        }
    }
}
